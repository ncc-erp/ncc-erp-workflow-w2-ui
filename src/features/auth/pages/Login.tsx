import {
  Box,
  Image,
  VStack,
  Heading,
  Button,
  Grid,
  Icon,
} from '@chakra-ui/react';
import { useForm } from 'react-hook-form';
import LoginBackground from 'assets/images/login_background.jpg';
import Logo from 'assets/images/w2_logo.png';
import { PasswordField } from 'common/components/PasswordField';
import { LoginExternalParams, LoginParams } from 'models/user';
import { FcGoogle } from 'react-icons/fc';
import { TextField } from 'common/components/TextField';
import { useLogin, useLoginExternal } from 'api/apiHooks/userHooks';
import { LocalStorageKeys, LoginStatus } from 'common/enums';
import { useNavigate } from 'react-router-dom';
import { toast } from 'common/components/StandaloneToast';
import { userManager } from 'services/authService';
import { setItem } from 'utils';

const initialLoginParams: LoginParams = {
  userNameOrEmailAddress: '',
  password: '',
  rememberMe: true,
};

const Login = () => {
  const navigate = useNavigate();
  const { mutateAsync: loginMutate, isLoading: isLoginLoading } = useLogin();
  const {
    mutateAsync: loginExternalMutate,
    isLoading: isLoginExternalLoading,
  } = useLoginExternal();

  const {
    handleSubmit,
    register,
    formState: { errors },
  } = useForm<LoginParams>({ defaultValues: initialLoginParams });

  const onLogin = async ({
    userNameOrEmailAddress,
    password,
    rememberMe,
  }: LoginParams) => {
    const { result } = await loginMutate({
      rememberMe,
      userNameOrEmailAddress: userNameOrEmailAddress.trim(),
      password: password.trim(),
    });

    if (result === LoginStatus.success) {
      navigate('/');
      return;
    }

    toast({
      title: 'Login Failed!',
      description: 'Invalid username or password!',
      status: 'error',
    });
  };

  const onLoginExternal = async ({
    provider,
    idToken,
  }: LoginExternalParams) => {
    const { token } = await loginExternalMutate({
      provider,
      idToken,
    });
    if (token) {
      setItem(LocalStorageKeys.accessToken, token);
    }
  };

  const handleLogin = async () => {
    const currentUser = await userManager.signinPopup();
    await onLoginExternal({
      provider: 'Google',
      idToken: currentUser.id_token,
    });
    window.location.href = '/';
  };

  return (
    <Grid
      templateColumns={{ base: 'auto', xl: '780px 1fr' }}
      backgroundImage={`url(${LoginBackground})`}
      backgroundPosition="60% 50%"
    >
      <VStack
        p={{ base: '14px', md: '38px 44px' }}
        h="100vh"
        backgroundColor="white"
        alignItems="left"
      >
        <Box width="full">
          <Image h="36px" src={Logo} />
        </Box>
        <VStack
          flex={1}
          mx="auto"
          w={{ base: 'full', md: '380px' }}
          justifyContent="center"
          spacing="20px"
        >
          <Heading as="h1" size="md" textAlign="left" w="full" fontWeight={700}>
            Sign In
          </Heading>
          <form style={{ width: '100%' }} onSubmit={handleSubmit(onLogin)}>
            <VStack spacing="14px">
              <TextField
                h="50px"
                placeholder="Username or email address"
                fontSize="sm"
                error={errors.userNameOrEmailAddress?.message}
                {...register('userNameOrEmailAddress', {
                  required: 'The Username or email address field is required!',
                })}
              />
              <PasswordField
                h="50px"
                placeholder="Password"
                fontSize="sm"
                error={errors.password?.message}
                {...register('password', {
                  required: 'The password field is required!',
                })}
                iconsProps={{
                  w: '18px',
                  h: '18px',
                }}
                buttonProps={{
                  mr: '10px',
                }}
              />
              <Button
                mt="14px"
                h="50px"
                type="submit"
                isLoading={isLoginLoading}
                colorScheme="blackButton"
                w="full"
                textColor="white"
              >
                Sign in
              </Button>
            </VStack>
          </form>
          <Button
            isLoading={isLoginExternalLoading}
            onClick={handleLogin}
            w="full"
            h="50px"
          >
            <Icon as={FcGoogle} mr="16px" filter="" fontSize="24px" />
            Sign in with Google
          </Button>
        </VStack>
      </VStack>
    </Grid>
  );
};

export default Login;
