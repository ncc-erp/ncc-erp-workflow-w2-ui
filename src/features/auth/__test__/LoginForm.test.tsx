import { screen, render } from '@testing-library/react';
import Login from '../pages/Login';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

jest.mock('../../../api/axiosInstant', () => ({
  VITE_API_BASE_URL: '/api',
}));

jest.mock('../../../services/authService', () => ({
  VITE_AUTHORITY_URL: '/api',
  VITE_OAUTH_CLIENT_ID:
    '797442917082-hihd7e20h4bef4oh0fbr7b8lml27ki7n.apps.googleusercontent.com',
  VITE_GOOGLE_LOGIN_REDIRECT: 'http://localhost:4200',
}));

jest.mock('../../../api/axiosInstant', () => ({
  useAxios: jest.fn(),
}));

describe('Login Page', () => {
  let queryClient: QueryClient;
  beforeAll(() => {
    queryClient = new QueryClient();
  });

  test('Check Display form Log in', async () => {
    render(
      <QueryClientProvider client={queryClient}>
        <Login />
      </QueryClientProvider>
    );
    const form = screen.getByTestId('login-form');
    expect(form).toBeInTheDocument();
  });
});
