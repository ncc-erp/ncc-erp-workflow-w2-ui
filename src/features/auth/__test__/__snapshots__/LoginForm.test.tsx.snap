// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Check Display form Log in 1`] = `
<div>
  <div
    class="css-c8j7f1"
  >
    <div
      class="chakra-stack css-1ftleaf"
    >
      <div
        class="css-13rbvqk"
      >
        <img
          class="chakra-image css-1fzkik5"
          src=""
        />
      </div>
      <div
        class="chakra-stack css-6jsom3"
      >
        <h1
          class="chakra-heading css-7esaea"
        >
          Sign In
        </h1>
        <form
          data-testid="login-form"
          style="width: 100%;"
        >
          <div
            class="chakra-stack css-axmyh0"
          >
            <div
              class="chakra-form-control css-0"
              role="group"
            >
              <div
                data-testid="userNameOrEmailAddress"
              >
                <input
                  class="chakra-input css-1ksr9b"
                  id="field-:r0:"
                  name="userNameOrEmailAddress"
                  placeholder="Username or email address"
                  type="text"
                />
              </div>
            </div>
            <div
              class="chakra-form-control css-0"
              role="group"
            >
              <div>
                <div
                  class="chakra-input__group css-bx0blc"
                  data-group="true"
                >
                  <input
                    class="chakra-input css-1ksr9b"
                    id="field-:r1:"
                    name="password"
                    placeholder="Password"
                    type="password"
                  />
                  <div
                    class="chakra-input__right-element css-1aezdzf"
                  >
                    <button
                      aria-label=""
                      class="chakra-button css-d90ir5"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="chakra-icon css-16sl94d"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 512 512"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M432 448a15.92 15.92 0 01-11.31-4.69l-352-352a16 16 0 0122.62-22.62l352 352A16 16 0 01432 448zm-176.34-64c-41.49 0-81.5-12.28-118.92-36.5-34.07-22-64.74-53.51-88.7-91v-.08c19.94-28.57 41.78-52.73 65.24-72.21a2 2 0 00.14-2.94L93.5 161.38a2 2 0 00-2.71-.12c-24.92 21-48.05 46.76-69.08 76.92a31.92 31.92 0 00-.64 35.54c26.41 41.33 60.4 76.14 98.28 100.65C162 402 207.9 416 255.66 416a239.13 239.13 0 0075.8-12.58 2 2 0 00.77-3.31l-21.58-21.58a4 4 0 00-3.83-1 204.8 204.8 0 01-51.16 6.47zm235.18-145.4c-26.46-40.92-60.79-75.68-99.27-100.53C349 110.55 302 96 255.66 96a227.34 227.34 0 00-74.89 12.83 2 2 0 00-.75 3.31l21.55 21.55a4 4 0 003.88 1 192.82 192.82 0 0150.21-6.69c40.69 0 80.58 12.43 118.55 37 34.71 22.4 65.74 53.88 89.76 91a.13.13 0 010 .16 310.72 310.72 0 01-64.12 72.73 2 2 0 00-.15 2.95l19.9 19.89a2 2 0 002.7.13 343.49 343.49 0 0068.64-78.48 32.2 32.2 0 00-.1-34.78z"
                        />
                        <path
                          d="M256 160a95.88 95.88 0 00-21.37 2.4 2 2 0 00-1 3.38l112.59 112.56a2 2 0 003.38-1A96 96 0 00256 160zm-90.22 73.66a2 2 0 00-3.38 1 96 96 0 00115 115 2 2 0 001-3.38z"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <button
              class="chakra-button css-1jntfhr"
              type="submit"
            >
              Sign in
            </button>
          </div>
        </form>
        <button
          class="chakra-button css-16as47f"
          type="button"
        >
          <svg
            class="chakra-icon css-1532g4u"
            enable-background="new 0 0 48 48"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            version="1.1"
            viewBox="0 0 48 48"
            width="1em"
            x="0px"
            xmlns="http://www.w3.org/2000/svg"
            y="0px"
          >
            <path
              d="M43.611,20.083H42V20H24v8h11.303c-1.649,4.657-6.08,8-11.303,8c-6.627,0-12-5.373-12-12
	c0-6.627,5.373-12,12-12c3.059,0,5.842,1.154,7.961,3.039l5.657-5.657C34.046,6.053,29.268,4,24,4C12.955,4,4,12.955,4,24
	c0,11.045,8.955,20,20,20c11.045,0,20-8.955,20-20C44,22.659,43.862,21.35,43.611,20.083z"
              fill="#FFC107"
            />
            <path
              d="M6.306,14.691l6.571,4.819C14.655,15.108,18.961,12,24,12c3.059,0,5.842,1.154,7.961,3.039l5.657-5.657
	C34.046,6.053,29.268,4,24,4C16.318,4,9.656,8.337,6.306,14.691z"
              fill="#FF3D00"
            />
            <path
              d="M24,44c5.166,0,9.86-1.977,13.409-5.192l-6.19-5.238C29.211,35.091,26.715,36,24,36
	c-5.202,0-9.619-3.317-11.283-7.946l-6.522,5.025C9.505,39.556,16.227,44,24,44z"
              fill="#4CAF50"
            />
            <path
              d="M43.611,20.083H42V20H24v8h11.303c-0.792,2.237-2.231,4.166-4.087,5.571
	c0.001-0.001,0.002-0.001,0.003-0.002l6.19,5.238C36.971,39.205,44,34,44,24C44,22.659,43.862,21.35,43.611,20.083z"
              fill="#1976D2"
            />
          </svg>
          Sign in with Google
        </button>
        <button
          class="chakra-button css-1q34lwh"
          type="button"
        >
          <img
            src=""
            width="24"
          />
          Sign in with Mezon
        </button>
      </div>
    </div>
  </div>
</div>
`;
